golem/task/rpc.py

```
@@ -537,13 +533,47 @@ def restart_task(self, task_id: str, force: bool = False) \
         self.task_manager.put_task_in_restarted_state(task_id)
         return new_task.header.task_id, None
 
+    @rpc_utils.expose('comp.task.subtasks.frame.restart')
+    @safe_run(
+        lambda e, _self, task_id, frame: logger.error(
+            'Frame restart failed. e=%r, task_id=%r, frame=%r',
+            e, task_id, frame
+        )
+    )
+    def restart_frame_subtasks(
+            self,
+            task_id: str,
+            frame: int
+    ):
+        logger.debug('restart_frame_subtasks. task_id=%r, frame=%r',
+                     task_id, frame)
+
+        frame_subtasks: typing.Dict[str, SubtaskState] =\
+            self.task_manager.get_frame_subtasks(task_id, frame)
+
+        if not frame_subtasks:
+            logger.error('Frame restart failed, frame has no subtasks.'
+                         'task_id=%r, frame=%r', task_id, frame)
+            return
```

IMO we should raise an exception here, so that the other side (CLI or GUI) will know the cause of failure

---

Actually, if look into `get_frame_subtasks` you can see that `None` is returned also when task with the given ID is not found. This relates to the next except clause. If `frame_subtasks` is not `None` and getting `task_state` raises a `KeyError` it means that something bad has happened and `taskmanager` is in an invalid state.

---

Got it, in that case I'm going to drop the next try-except and keep the `@run_safe` decorator for catching any unexpected errors.